core/src/SYCL/Kokkos_SYCL_ParallelFor_Range.hpp

```diff
@@ -89,12 +94,33 @@ class Kokkos::Impl::ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>,
 #else
       (void)memcpy_event;
 #endif
+
       if (policy.chunk_size() <= 1) {
+#ifdef SYCL_EXT_ONEAPI_AUTO_LOCAL_RANGE
+        const auto actual_range = policy.end() - policy.begin();
+        FunctorWrapperRangePolicyParallelForCustom<Functor, Policy> f{
+            policy.begin(), functor, actual_range};
+        // Round the actual range up to the closest power of two not exceeding
+        // the maximum workgroup size
+        const auto max_wgroup_size =
+            q.get_device().get_info<sycl::info::device::max_work_group_size>();
+        const auto wgroup_size_multiple = Kokkos::bit_floor(
+            std::min<std::size_t>(max_wgroup_size, actual_range));
+
+        const auto launch_range = (actual_range + wgroup_size_multiple - 1) /
+                                  wgroup_size_multiple * wgroup_size_multiple;
```

Comment:
So we can't rely on the implementation to adjust the size of the range?

Reply:
The implementation will choose a workgroup size that divides the range which means 1 if the number of work items is a prime, see https://github.com/kokkos/kokkos/issues/6745#issuecomment-1912734891. We assume that the selected subgroup size will always be a power-of-two and so we increase the range accordingly and guard in the kernel for the actual range.